casper/src/main/resources/PoS.rhox

```diff
@@ -63,626 +66,646 @@ in {
+                              for (@quarantinedValidators <- quarantinedValidatorsCh) {
+                                @ListOps!("map", quarantinedValidators, *fst, *validatorsToWithdrawListCh) |
+                                for (@validatorsToWithdrawList <- validatorsToWithdrawListCh) {
+                                  @ListOps!("map", validatorsToWithdrawList, *computePay, *paymentsCh) |
+                                  contract computePay(@pk, resultCh) = {
+                                    resultCh!((pk, allBonds.get(pk) + rewards.getOrElse(pk, 0)))
                                   } |
-
-                                  contract notWithdrawn(@(pk, _), resultCh) = { resultCh!(not validatorsToWithdrawSet.contains(pk))}
+                                  match validatorsToWithdrawList.toSet() {
+                                    validatorsToWithdrawSet => {
+                                      @ListOps!("filter", allBonds.toList(), *notWithdrawn, *newBondsListCh) |
+                                      @ListOps!("filter", withdrawers.toList(), *notWithdrawn, *newWithdrawersListCh) |
+                                      @ListOps!("filter", rewards.toList(), *notWithdrawn, *newRewardsListCh) |
+                                      for (@newWithdrawersList <- newWithdrawersListCh;
+                                           @newBondsList       <- newBondsListCh;
+                                           @newRewardsList     <- newRewardsListCh;
+                                           @payments           <- paymentsCh) {
+                                        returnCh!((payments, newBondsList.toMap(), newWithdrawersList.toMap(), newRewardsList.toMap()))
+                                      } |
+                                      // Checks that the validator is not a withdrawer.
+                                      contract notWithdrawn(@(pk, _), resultCh) = {
+                                        resultCh!(not validatorsToWithdrawSet.contains(pk))
+                                      }
+                                    }
+                                  }
                                 }
+                              } |
+                              // Check whether quarantine period is finished.
+                              contract isQuarantineFinished(@(_, blockNumber), resultCh) = {
+                                resultCh!(blockNumber >= currentBlockNumber)
                               }
                             }
-                          } |
-
-                          contract isQuarantineFinished(@(_, blockNumber), resultCh) = {
-                            resultCh!(blockNumber >= currentBlockNumber)
                           }
                         }
+                      } else {
+                        ackCh!((false, "Invalid system auth token"))
                       }
```

Comment:
When successful `closeBlock` will return just `Nil` (line 477 send on `stateUpdateCh`). Maybe it makes sense to return the same type (Either). It's similar in "slash" which returns Boolean if successful.